java/generate-offline-map/src/main/java/com/esri/arcgisruntime/generateofflinemap/MainActivity.java

```diff
@@ -0,0 +1,246 @@
+    mDownloadArea.setSymbol(simpleLineSymbol);
+
+    // update the download area box whenever the viewpoint changes
+    mMapView.addViewpointChangedListener(viewpointChangedEvent -> {
+      if (map.getLoadStatus() == LoadStatus.LOADED) {
+        // upper left corner of the area to take offline
+        android.graphics.Point minScreenPoint = new android.graphics.Point(200, 200);
+        // lower right corner of the downloaded area
+        android.graphics.Point maxScreenPoint = new android.graphics.Point(mMapView.getWidth() - 200,
+            mMapView.getHeight() - 200);
+        // convert screen points to map points
+        Point minPoint = mMapView.screenToLocation(minScreenPoint);
+        Point maxPoint = mMapView.screenToLocation(maxScreenPoint);
+        // use the points to define and return an envelope
+        if (minPoint != null && maxPoint != null) {
+          Envelope envelope = new Envelope(minPoint, maxPoint);
+          mDownloadArea.setGeometry(envelope);
+        }
+      }
+    });
+  }
+
+  /**
+   * Use the generate offline map job to generate an offline map.
+   */
+  private void generateOfflineMap() {
+
+    // create a progress dialog to show download progress
+    ProgressDialog progressDialog = new ProgressDialog(this);
+    progressDialog.setTitle("Generate Offline Map Job");
+    progressDialog.setMessage("Taking map offline...");
+    progressDialog.setProgressStyle(ProgressDialog.STYLE_HORIZONTAL);
+    progressDialog.setIndeterminate(false);
+    progressDialog.setProgress(0);
+
+    // when the button is clicked, start the offline map task job
+    mTakeMapOfflineButton.setOnClickListener(v -> {
+      progressDialog.show();
+
+      // delete any offline map already in the cache
+      String tempDirectoryPath = getCacheDir() + File.separator + "offlineMap";
+      deleteDirectory(new File(tempDirectoryPath));
+
+      // specify the extent, min scale, and max scale as parameters
+      double minScale = mMapView.getMapScale();
+      double maxScale = mMapView.getMap().getMaxScale();
+      // minScale must always be larger than maxScale
+      if (minScale <= maxScale) {
+        minScale = maxScale + 1;
+      }
+      GenerateOfflineMapParameters generateOfflineMapParameters = new GenerateOfflineMapParameters(
+          mDownloadArea.getGeometry(), minScale,
+          maxScale);
+
+      // create an offline map offlineMapTask with the map
+      OfflineMapTask offlineMapTask = new OfflineMapTask(mMapView.getMap());
+
+      // create an offline map job with the download directory path and parameters and start the job
+      GenerateOfflineMapJob job = offlineMapTask.generateOfflineMap(generateOfflineMapParameters, tempDirectoryPath);
+
+      // replace the current map with the result offline map when the job finishes
+      job.addJobDoneListener(() -> {
+        if (job.getStatus() == Job.Status.SUCCEEDED) {
```

Comment:
You might want to add an else-statement if the job does not succeed

Reply:
It's already got one.